src/lib/transition_router/pipe_guard.ml

```diff
@@ -0,0 +1,71 @@
+open Core
+open Async
+open Pipe_lib
+
+module Make (Data : sig
+  type t
+
+  val to_yojson : t -> Yojson.Safe.t
+end) =
+struct
+  type writer =
+    ( Data.t
+    , Strict_pipe.drop_head Strict_pipe.buffered
+    , unit )
+    Strict_pipe.Writer.t
+
+  type identified_writer = { writer : writer; id : int }
+
+  type control_msg =
+    | ClosePipe of
+        { result : [ `Closed_already_or_replaced | `Accepted ] Ivar.t
+        ; id : int
+        }
+    | ReplacePipe of { writer : writer; id_returned : int Ivar.t }
+
+  type state = { writer : identified_writer option; next_id : int }
+
+  module Guard = Actor.Make (Data)
+
+  let data_handler ~(state : state) ~(message : Data.t) =
+    let rec wait_till_writer_ready () =
+      match state.writer with
+      | None ->
+          let%bind () = Async.Scheduler.yield () in
+          wait_till_writer_ready ()
+      | Some writer_with_id ->
+          Deferred.return writer_with_id
+    in
+    let%map writer_with_id = wait_till_writer_ready () in
+    Strict_pipe.Writer.write writer_with_id.writer message ;
+    Guard.Next state
+
+  let control_handler ~(state : state) ~(message : control_msg) =
```

Comment:
Note: it doesn't matter the order we handle the signals inside an actor, since no contention to write will ever happen -- all handlers are ran in sequential.